### ISSUE TITLE
Core dumps fill the host's disk

### ISSUE THREAD

**Issue by AnthonyMastrean** [dotnet/dotnet-docker#1215]
A long-running .NET Core or ASP.NET Core application will encounter unhandled exceptions, abort, and create core dumps until the "disk" is full. Sure, these will happen at unknown or non-regular frequencies, but it's possible.

The example I provide accelerates this lifecycle to make the point 😄 

## Steps to reproduce the issue

Create a .NET Core console application or ASP.NET Core application and replace the contents of the `Main` method with `throw new Exception()`, like in https://github.com/dotnet/coreclr/issues/17929#issue-321559773. Build that application into an image, per the appropriate [sample](https://github.com/dotnet/dotnet-docker/tree/master/samples).

Run the container with the "always" restart policy (which, I think, is a reasonable policy for production and perhaps part of what makes this issue different than the others).

```
$ docker run -d --restart always example
```

## Expected behavior

Long-running .NET Core and ASP.NET Core apps shouldn't fill the disk with core dumps until the host system crashes or fails or whatever.

## Actual behavior

The system creates core dumps until the disk is full. Here's an snippet from a service running and crashing on CentOS 7.6.

```
$ cat /proc/sys/kernel/core_pattern 
|/usr/libexec/abrt-hook-ccpp %s %c %p %u %g %t e %P %I %h
```
```
$ ls -lah
total 6.8G
drwxr-xr-x. 2 root root 4.0K Jul  2 18:50 .
drwxr-xr-x. 3 root root   16 Jun 28 18:15 ..
-rw-------. 1 root root 153M Jul  2 17:20 core.10034
-rw-------. 1 root root 153M Jul  2 17:49 core.10097
-rw-------. 1 root root 153M Jul  2 17:23 core.10263
-rw-------. 1 root root 153M Jul  2 18:14 core.10553
-rw-------. 1 root root 153M Jul  2 17:15 core.10825
-rw-------. 1 root root 153M Jul  2 18:34 core.10899
-rw-------. 1 root root 153M Jul  2 17:52 core.10915
-rw-------. 1 root root 153M Jul  2 17:18 core.11197
-rw-------. 1 root root 153M Jul  2 18:39 core.11438
```

## Additional information

It's clear from https://github.com/dotnet/coreclr/issues/17929 and https://github.com/aspnet/AspNetCore/issues/5920 that the issue is _currently_ considered `wontfix`. Forget the esoteric technical argument(s) about aborting, dumping, or returning exit codes... This behavior has the capability to adversely effect the host system and other services on the host. That's bad!

My options are all sub-optimal, let's review (in order of broadest to narrowest impact)...

 1. The `core_pattern` is inherited from the host. Disabling dumps on the host impacts the entire host and all other services running on the host.
 1. You can modify the Docker daemon's conf (`daemon.json`) or use the CLI argument `--ulimit core=0` to disable core dumps. But, this effects all containers on the host (even non-.NET Core containers).
 1. The Docker `run` command supports the same `--ulimit` argument (so does Compose). But, now I have to _remember_ to run my .NET Core services this way or litter my Compose manifest(s) with kernel parameters.
 1. It seems effective to add a `RUN ulimit -c 0` into the Dockerfile. That means additional lines that have _nothing_ really to do with your application. Or your team would need to take this into the base image (which seems unlikely).
 1. The core files appear to be created in the application directory (and I'm not sure I can change that, again, without effecting the behavior across the entire host), so I can't put them on a host VOLUME.

It's clear that core dumps in Docker, generally, are a pretty gnarly problem: https://github.com/moby/moby/issues/11740 https://github.com/moby/moby/issues/19289

## Output of `docker version`

(any)

## Output of `docker info`

(any)

**Comment by MichaelSimons** [comment]
@AnthonyMastrean - can you clarify something about your CentOS environment?  Are the dumps triggered within the container being written to the host or the container?  If they are written to the host how is this accomplished?  Docker volume mount? other?

> Long-running .NET Core and ASP.NET Core apps shouldn't fill the disk with core dumps until the host system crashes or fails or whatever.

In this scenario, I have troubles considering the app to be long running in a way.  It is the auto restart nature that makes it "long running".  Wouldn't you want monitoring in place to know this is happening so that the underlying source that is triggering the restart can be addressed?  I am apprehensive about the `always` option, on-failure[:max-retries] is more attractive to me.

> It seems effective to add a RUN ulimit -c 0 into the Dockerfile. That means additional lines that have nothing really to do with your application. Or your team would need to take this into the base image (which seems unlikely).

I agree, I don't see this as something appropriate to put into the base image.

> It's clear that core dumps in Docker, generally, are a pretty gnarly problem: moby/moby#11740 moby/moby#19289

The proposal in moby/moby#19289 seems beneficial for scenarios like this.

Given the behavior of .NET Core and ASP.NET Core along with the Docker behavior and limitations, it is not clear to me what changes could and should be made to the .NET Core images .

**Comment by AnthonyMastrean** [comment]
This is my understanding of the file location:

> The core files appear to be created in the application directory (and I'm not sure I can change that, again, without effecting the behavior across the entire host), so I can't put them on a host VOLUME.

I certainly _accelerated_ the problem `restart: always`, but that was kinda the point... core files are uncontrollably written to the container's filesystem. I guess I'm just adding my perspective to the overall argument...

 A) .NET Core should stop core dumping in the case of unhandled exceptions, because
 B) Docker doesn't handle core dumps particularly gracefully

There doesn't seem to be strong support for A as a short-term or long-term solution (B would still exist, anyway). Although, the Docker team might be making progress on B.

Feel free to close, if we're all just waiting on B.

**Comment by MichaelSimons** [comment]
Closing per discussion.
